### PR TITLE
feat(BlogPost): add post about private Cdn logs

### PR DESCRIPTION
**Pull request by arthurzinck**
# Why
It add a Blog Post titled : `How to ingest 400GB of logs per hour ?`
It talks about the how we collect and make our logs available to all Bedrock teams.
I wanted it to be not to technical, more destined to cloud architects.

# How
- adding me as an author
- adding my avatar
- adding blog post
- adding image for the blog post

### CODE REVIEW

File: _posts/2022-08-02-private-cdn-logs.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
+---
+layout: post
+title: "How to ingest 400GB of logs per hour ?"
```
Comment (christian-vdz) [inline _posts/2022-08-02-private-cdn-logs.md:3]:
```suggestion
title: "How to ingest 400GB of logs per hour?"
```
2 écoles

File: _posts/2022-08-02-private-cdn-logs.md [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+---
+Bedrock Streaming is a company that sells a white label streaming and live platform. Our customers are media groups, TV channels, and streaming companies. Our goal is to deliver a state of the art streaming  platform to our customers.
+
+To achieve this goal, we have our own Content Delivery Network (CDN), made of several bare metal servers racked in our Data Centers. Those servers run Nginx and are designed to output hundreds of Gbps (several tens of Pb per month) to end-users. We use them to cache video content at our infrastructure’s edge. 
+
+This increases efficiency 96 times out of 100, as video traffic doesn’t have to flow all the way through our infrastructure and improves user experience as it serves video faster. Also it diminishes the cost of our Video On Demand (VOD) infrastructure as we need less servers in VOD Stack.
+
+This in-turn increases end-users (clients of our customers) satisfaction with the service.
+
+## Who needs to ingest 400GB of logs per hour anyway ?
```
Comment (christian-vdz) [inline _posts/2022-08-02-private-cdn-logs.md:20]:
```suggestion
## Who needs to ingest 400GB of logs per hour anyway?
```

File: _posts/2022-08-02-private-cdn-logs.md [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+title: "How to ingest 400GB of logs per hour ?"
+description: "We wanted to exploit our CDN logs as they contains many valuable information."
+author: arthurzinck
+tags: [onprem, cdn, logs, aws, cloud, nginx, terraform, lambda, s3, glue, athena]
+color: rgb(251,87,66)
+# thumbnail: ""
+language: en
+comments: true
+---
+Bedrock Streaming is a company that sells a white label streaming and live platform. Our customers are media groups, TV channels, and streaming companies. Our goal is to deliver a state of the art streaming  platform to our customers.
```
Comment (christian-vdz) [inline _posts/2022-08-02-private-cdn-logs.md:12]:
```suggestion
Bedrock Streaming is a company that sells a white labeled streaming and live platform. Our customers are media groups and streaming companies. Our goal is to deliver them a state of the art streaming platform.
```
TV channels are part of media groups

File: _posts/2022-08-02-private-cdn-logs.md [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+To achieve this goal, we have our own Content Delivery Network (CDN), made of several bare metal servers racked in our Data Centers. Those servers run Nginx and are designed to output hundreds of Gbps (several tens of Pb per month) to end-users. We use them to cache video content at our infrastructure’s edge.
```
Comment (christian-vdz) [inline _posts/2022-08-02-private-cdn-logs.md:14]:
```suggestion
To achieve this goal, we have our own Content Delivery Network (CDN), made of several bare metal servers racked in our Data Centers. Those servers cache video content at edge running Nginx and are designed to output hundreds of Gbps (several tens of Pb per month) to end-users.
```

File: _posts/2022-08-02-private-cdn-logs.md [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+color: rgb(251,87,66)
+# thumbnail: ""
+language: en
+comments: true
+---
+Bedrock Streaming is a company that sells a white label streaming and live platform. Our customers are media groups, TV channels, and streaming companies. Our goal is to deliver a state of the art streaming  platform to our customers.
+
+To achieve this goal, we have our own Content Delivery Network (CDN), made of several bare metal servers racked in our Data Centers. Those servers run Nginx and are designed to output hundreds of Gbps (several tens of Pb per month) to end-users. We use them to cache video content at our infrastructure’s edge. 
+
+This increases efficiency 96 times out of 100, as video traffic doesn’t have to flow all the way through our infrastructure and improves user experience as it serves video faster. Also it diminishes the cost of our Video On Demand (VOD) infrastructure as we need less servers in VOD Stack.
```
Comment (christian-vdz) [inline _posts/2022-08-02-private-cdn-logs.md:16]: What efficiency?

File: _posts/2022-08-02-private-cdn-logs.md [hunk 6]
```diff
@@ -0,0 +1,103 @@
+---
+layout: post
+title: "How to ingest 400GB of logs per hour ?"
+description: "We wanted to exploit our CDN logs as they contains many valuable information."
```
Comment (vgallissot) [inline _posts/2022-08-02-private-cdn-logs.md:4]:
```suggestion
description: "We wanted to exploit our CDN logs as they contain many valuable information."
```

File: _posts/2022-08-02-private-cdn-logs.md [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+This increases efficiency 96 times out of 100, as video traffic doesn’t have to flow all the way through our infrastructure and improves user experience as it serves video faster. Also it diminishes the cost of our Video On Demand (VOD) infrastructure as we need less servers in VOD Stack.
+
+This in-turn increases end-users (clients of our customers) satisfaction with the service.
+
+## Who needs to ingest 400GB of logs per hour anyway ?
+Every time someone watches a video, it generates traffic on our CDN, resulting in a lot of access logs. Without filtering, it averages to 400GB uncompressed logs per hour.
+
+This is why, at first, we chose to not log 2XX or 3XX HTTP codes. We had too many of them and we considered them not as worth it as 4XX and 5XX. The 4XX and 5XX can be especially useful for debugging a particular situation or from a broader perspective, improving the user experience.
+
+This was the kind of Nginx configuration we had deployed :
```
Comment (christian-vdz) [inline _posts/2022-08-02-private-cdn-logs.md:25]:
```suggestion
This was the kind of Nginx configuration we deployed :
```

**Review by vgallissot — APPROVED** [review]

File: _posts/2022-08-02-private-cdn-logs.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+map $status $loggable {
+    ~^[23]  0;
+    default 1;
+}
+access_log /path/to/access.log combined if=$loggable;
+```
+
+## Giving autonomy for all teams on logs
+
+At the end of 2021, the finance team approached us with a challenge: how to bill our customers based on their end-users CDN usage?
```
Comment (christian-vdz) [inline _posts/2022-08-02-private-cdn-logs.md:36]:
```suggestion
At the end of 2021, the finance team approached us with a challenge: how to bill our customers based on their end-users CDN usage?  
```

File: _posts/2022-08-02-private-cdn-logs.md [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+This was in fact a need we already anticipated, we tried the nginx module [Traffic_accounting](https://www.nginx.com/resources/wiki/modules/traffic_accounting/) but it did not satisfy us fully. This module calculates and exposes metrics on-the-fly, which is CPU and memory intensive, especially above 50Gbps of traffic per server.
+
+We also had another objective that wasn’t addressed with the nginx module. We needed to give autonomy to QA, Video, Data and Finance teams. We wanted to allow them to use CDN logs when they needed without having to ask for it, and ideally in a practical and unified way. 
+
+The company philosophy states that we are user obsessed and that we do not finger point. We work as a team to offer the best user experience, this is why we make all our logs available to all teams. We didn’t come around to do it for the CDN as the volume of logs was too much of a constraint.
+
+## Technical Solution
+
+At bedrock we like to keep things simple. We think our CDN main mission is to serve video as efficiently as possible. Our CDN’s servers can’t keep PetaBytes of logs on their disks. This is why we chose to output logs to Amazon S3.
```
Comment (christian-vdz) [inline _posts/2022-08-02-private-cdn-logs.md:45]:
```suggestion
At Bedrock we like to keep things simple. We think our CDN main mission is to serve video as efficiently as possible. Our CDN’s servers can’t keep PetaBytes of logs on their disks. This is why we chose to output logs to Amazon S3.
```

File: _posts/2022-08-02-private-cdn-logs.md [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+ 
+To send logs from our CDN servers to Amazon S3 bucket, we had many options but chose to test two approaches: [Fluentd](https://www.fluentd.org/) and [Vector](https://vector.dev/). Fluentd is the legacy one, and Vector the new rusty one.
+
+
+After a quick evaluation we decided to go with [Vector as it seemed more memory efficient](https://medium.com/ibm-cloud/log-collectors-performance-benchmarking-8c5218a08fea) and output more Logs Per Second under heavy load than Fluentd.
+
+
+![Log per second](/images/posts/2022-08-10-privateCdnLogs/image4.png)
+
+source: [Who is the winner — Comparing Vector, Fluent Bit, Fluentd performance from Ajay Gupta](https://medium.com/ibm-cloud/log-collectors-performance-benchmarking-8c5218a08fea)
```
Comment (christian-vdz) [inline _posts/2022-08-02-private-cdn-logs.md:61]:
```suggestion
<center><i>source: [Who is the winner — Comparing Vector, Fluent Bit, Fluentd performance from Ajay Gupta](https://medium.com/ibm-cloud/log-collectors-performance-benchmarking-8c5218a08fea)</i></center>
```

File: _posts/2022-08-02-private-cdn-logs.md [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+
+
+We have Nginx and Vector installed on the CDN servers. Nginx now outputs all the access logs to a file. Vector reads the file, compresses logs to GZIP format and every 10Mb sends the logs to S3. Nginx may generate at peak 600GB of logs; we only send 10GB.
+
+Those logs are then locally cleaned by Logrotate.
+
+### Storing logs: S3
+We have chosen to store logs on an S3 bucket. We figured it was the most scalable and time efficient. S3 buckets can grow to PetaBytes easily. It is a few terraform lines away, this is convenient as we handle all our infrastructure with Terraform.
```
Comment (christian-vdz) [inline _posts/2022-08-02-private-cdn-logs.md:69]:
```suggestion
We chose to store logs on an S3 bucket. We figured it was the most scalable and time efficient. S3 buckets can grow to PetaBytes easily. It is a few terraform lines away, this is convenient as we handle all our infrastructure with Terraform.
```

File: _posts/2022-08-02-private-cdn-logs.md [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+We configured our bucket to use several lifecycle policies. One to automatically clean logs after 365 days, another to remove incomplete uploads and another one to immediately remove files with a delete marker. Also we configured the storage class in *intelligent tiering mode* to store logs according to their access frequency.
+
+This will permit us to diminish the cost of our S3 bucket and not have an ever increasing S3 bill.
+
+### Partitioning logs on S3: Lambda stack
+
+Once logs are stored in S3 bucket, we need to classify and sort them in order to extract valuable intel. At Bedrock we already use a modified version of a lambda stack, that does just that. Originally designed for Cloudfront, we have been using it also for Fastly and now for our Private CDN. You can find the original version at [Aws Sample github](https://github.com/aws-samples/amazon-cloudfront-access-logs-queries).
```
Comment (christian-vdz) [inline _posts/2022-08-02-private-cdn-logs.md:77]:
```suggestion
Once logs are stored in S3 bucket, we need to classify and sort them in order to extract valuable intel. At Bedrock we already use a modified version of a lambda stack, that does just that. Originally designed for Cloudfront, we have been using it also for Fastly and now for our Private CDN. You can find the original version at [AWS Sample Github](https://github.com/aws-samples/amazon-cloudfront-access-logs-queries).
```

File: _posts/2022-08-02-private-cdn-logs.md [hunk 13]
```diff
@@ -0,0 +1,103 @@
+---
+layout: post
+title: "How to ingest 400GB of logs per hour ?"
+description: "We wanted to exploit our CDN logs as they contains many valuable information."
+author: arthurzinck
+tags: [onprem, cdn, logs, aws, cloud, nginx, terraform, lambda, s3, glue, athena]
```
Comment (christian-vdz) [inline _posts/2022-08-02-private-cdn-logs.md:6]:
```suggestion
tags: [onprem, cdn, logs, aws, cloud, nginx, vector, lambda, s3, glue, athena]
```
+/-

File: _posts/2022-08-02-private-cdn-logs.md (hunk 10 above)
Comment (arthurzinck) [reply _posts/2022-08-02-private-cdn-logs.md:61]: Let me check how i can do better on this

File: _posts/2022-08-02-private-cdn-logs.md (hunk 2 above)
Comment (arthurzinck) [reply _posts/2022-08-02-private-cdn-logs.md:20]: I prefer with the space

File: _posts/2022-08-02-private-cdn-logs.md (hunk 1 above)
Comment (arthurzinck) [reply _posts/2022-08-02-private-cdn-logs.md:3]: I prefer with the space

File: _posts/2022-08-02-private-cdn-logs.md (hunk 3 above)
Comment (arthurzinck) [reply _posts/2022-08-02-private-cdn-logs.md:12]: yes but media groups do not always have tv channel. I think it is good like that

File: _posts/2022-08-02-private-cdn-logs.md (hunk 5 above)
Comment (arthurzinck) [reply _posts/2022-08-02-private-cdn-logs.md:16]: I think it is clearly explained : ` as video traffic doesn’t have to flow all the way through our infrastructure and improves user experience as it serves video faster. Also it diminishes the cost of our Video On Demand (VOD) infrastructure as we need less servers in VOD Stack.`

File: _posts/2022-08-02-private-cdn-logs.md (hunk 4 above)
Comment (arthurzinck) [reply _posts/2022-08-02-private-cdn-logs.md:14]: I rather keep it in longer form as it I think easier to understand

File: _posts/2022-08-02-private-cdn-logs.md (hunk 3 above)
Comment (christian-vdz) [reply _posts/2022-08-02-private-cdn-logs.md:12]: fix spelling mistake at least :D

File: _posts/2022-08-02-private-cdn-logs.md [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+title: "How to ingest 400GB of logs per hour ?"
+description: "We wanted to exploit our CDN logs as they contain many valuable information."
+author: arthurzinck
+tags: [onprem, cdn, logs, aws, cloud, nginx, vector, lambda, s3, glue, athena]
+color: rgb(251,87,66)
+# thumbnail: ""
+language: en
+comments: true
+---
+Bedrock Streaming is a company that sells a white label streaming and live platform. Our customers are media groups, TV channels, and streaming companies. Our goal is to deliver a state of the art streaming  platform to our customers.
```
Comment (christian-vdz) [inline _posts/2022-08-02-private-cdn-logs.md:12]:
```suggestion
Bedrock Streaming is a company that sells a white labeled streaming and live platform. Our customers are media groups, TV channels, and streaming companies. Our goal is to deliver a state of the art streaming platform to our customers.
```

**Review by Slashgear — APPROVED** [review]
Blog post structure seems to be ok
Can't really judge the content otherwise :)

**Review by christian-vdz — APPROVED** [review]

File: _posts/2022-08-02-private-cdn-logs.md [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+title: "How to ingest 400GB of logs per hour ?"
+description: "We wanted to exploit our CDN logs as they contain many valuable information."
+author: arthurzinck
+tags: [onprem, cdn, logs, aws, cloud, nginx, vector, lambda, s3, glue, athena]
+color: rgb(251,87,66)
+# thumbnail: ""
+language: en
+comments: true
+---
+Bedrock Streaming is a company that sells a white labeled streaming and live platform. Our customers are media groups, TV channels, and streaming companies. Our goal is to deliver a state of the art streaming  platform to our customers.
```
Comment (pmartin) [inline _posts/2022-08-02-private-cdn-logs.md:12]: Cette première phrase est celle qui est reprise sur la page d'accueil. Et elle ne dit rien du contenu de l'article, elle ne donne pas envie de le lire :-(

Comment (arthurzinck) [reply _posts/2022-08-02-private-cdn-logs.md:13]: @pmartin what do you think of this override ? is it better ?

File: _posts/2022-08-02-private-cdn-logs.md [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+layout: post
+title: "How to ingest 400GB of logs per hour ?"
+description: "We wanted to exploit our CDN logs as they contain many valuable information."
+author: arthurzinck
+tags: [onprem, cdn, logs, aws, cloud, nginx, vector, lambda, s3, glue, athena]
+color: rgb(251,87,66)
+# thumbnail: ""
+language: en
+comments: true
+excerpt: "At Bedrock we have a CDN that outputs on average 400GB of uncompressed logs. In this article we will explain in an architectural point of vue how we managed to collect and extract value out of them."
```
Comment (pmartin) [inline _posts/2022-08-02-private-cdn-logs.md:11]:
```suggestion
excerpt: "At Bedrock, we have a CDN that outputs on average 400GB of uncompressed logs each day. In this article, we present the architecture we have setup to collect these logs and extract value from them."
```

File: _posts/2022-08-02-private-cdn-logs.md (hunk 15 above)
Comment (pmartin) [reply _posts/2022-08-02-private-cdn-logs.md:13]: Mieux, oui :-)
J'ai suggéré quelques retouches.

File: _posts/2022-08-02-private-cdn-logs.md [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+layout: post
+title: "How to ingest 400GB of logs per hour ?"
+description: "We wanted to exploit our CDN logs as they contain many valuable information."
+author: arthurzinck
+tags: [onprem, cdn, logs, aws, cloud, nginx, vector, lambda, s3, glue, athena]
+color: rgb(251,87,66)
+thumbnail: "/images/posts/2022-08-10-privateCdnLogs/main.png"
+language: en
+comments: true
+excerpt: "At Bedrock, we have a CDN that outputs on average 400GB of uncompressed logs per hour. In this article, we present the architecture we have setup to collect these logs and extract value from them."
```
Comment (mere-teresa) [inline _posts/2022-08-02-private-cdn-logs.md]: Grammatical: "we have set up"

File: _posts/2022-08-02-private-cdn-logs.md [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+---
+Bedrock Streaming is a company that sells a white labeled streaming and live platform. Our customers are media groups, TV channels, and streaming companies. Our goal is to deliver a state of the art streaming  platform to our customers.
+
+To achieve this goal, we have our own Content Delivery Network (CDN), made of several bare metal servers racked in our Data Centers. Those servers run Nginx and are designed to output hundreds of Gbps (several tens of Pb per month) to end-users. We use them to cache video content at our infrastructure’s edge. 
+
+This increases efficiency of the platform 96 times out of 100, as video traffic doesn’t have to flow all the way through our infrastructure and improves user experience as it serves video faster. Also it diminishes the cost of our Video On Demand (VOD) infrastructure as we need less servers in VOD Stack.
```
Comment (mere-teresa) [inline _posts/2022-08-02-private-cdn-logs.md]: Grammar: commas missing
after "our infrastructure,"
and after "Also,"

File: _posts/2022-08-02-private-cdn-logs.md [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+map $status $loggable {
+    ~^[23]  0;
+    default 1;
+}
+access_log /path/to/access.log combined if=$loggable;
+```
+
+## Giving autonomy for all teams on logs
+
+At the end of 2021, the finance team approached us with a challenge: how to bill our customers based on their end-users CDN usage ?
```
Comment (mere-teresa) [inline _posts/2022-08-02-private-cdn-logs.md]: Grammar: pas d'espace en anglais avant le point d'interrogation

File: _posts/2022-08-02-private-cdn-logs.md [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+The real benefit to using S3 is that you can easily plug it into Glue and Athena which allows you to request TeraBytes of data easily.
+
+![technical Solution](/images/posts/2022-08-10-privateCdnLogs/image1.png)
+
+### Sending logs to S3: Vector
+ 
+To send logs from our CDN servers to Amazon S3 bucket, we had many options but chose to test two approaches: [Fluentd](https://www.fluentd.org/) and [Vector](https://vector.dev/). Fluentd is the legacy one, and Vector the new rusty one.
+
+
+After a quick evaluation we decided to go with [Vector as it seemed more memory efficient](https://medium.com/ibm-cloud/log-collectors-performance-benchmarking-8c5218a08fea) and output more Logs Per Second under heavy load than Fluentd.
```
Comment (mere-teresa) [inline _posts/2022-08-02-private-cdn-logs.md]: Grammar: comma missing after "evaluation,"

File: _posts/2022-08-02-private-cdn-logs.md [hunk 21]
```diff
@@ -0,0 +1,100 @@
...
+### Storing logs: S3
+We chose to store logs on an S3 bucket. We figured it was the most scalable and time efficient. S3 buckets can grow to PetaBytes easily. It is a few terraform lines away, this is convenient as we handle all our infrastructure with Terraform.
+
+We configured our bucket to use several lifecycle policies. One to automatically clean logs after 365 days, another to remove incomplete uploads and another one to immediately remove files with a delete marker. Also we configured the storage class in *intelligent tiering mode* to store logs according to their access frequency.
+
+This will permit us to diminish the cost of our S3 bucket and not have an ever increasing S3 bill.
+
+### Partitioning logs on S3: Lambda stack
+
+Once logs are stored in S3 bucket, we need to classify and sort them in order to extract valuable intel. At Bedrock we already use a modified version of a lambda stack, that does just that. Originally designed for Cloudfront, we have been using it also for Fastly and now for our Private CDN. You can find the original version at [AWS Sample Github](https://github.com/aws-samples/amazon-cloudfront-access-logs-queries).
```
Comment (mere-teresa) [inline _posts/2022-08-02-private-cdn-logs.md]: Suggestion: remove "in order to" and write only "to" instead (see [Google Dev Styleguide](https://developers.google.com/style/word-list?hl=en#letter-i))

**Review by mere-teresa — DISMISSED** [review]
Please, use https://languagetool.org/ to spot the missing commas. (I pointed some but not all...)

**Pull request merged**